### ISSUE TITLE
OnnxRuntime gives different outputs on A100 v/s H100 GPU

### ISSUE THREAD

**Issue by akashkm99** [microsoft/onnxruntime#24027]
### Describe the issue

Hi ORT team,

We are observing significant discrepancies in the outputs when running inference on a standard BERT-like model using ONNX Runtime on A100 vs. H100 GPUs. Specifically, the hidden states corresponding to pad tokens differ between the two. Below is an example of the input and output:


|        | A100                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | H100                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|--------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Input  | { 'src_tokens': array([[31, 218232, 242834, 242834, 242834, 242834, 242834, 242834  242834, 242834, 242834, 242834, 242834, 242834, 242834, 242834]]), 'attention_mask': array([[1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=int32)}                                                                                                                                                                                                                                                                  | {'src_tokens': array([[31, 218232, 242834, 242834, 242834, 242834, 242834, 242834, 242834, 242834, 242834, 242834, 242834, 242834, 242834, 242834]]), 'attention_mask': array([[1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=int32)}                                                                                                                                                                                                                                                                      |
| Output | [[[ 1.9365234   0.09844971 -2.140625   ...  0.1394043  -1.3173828 -0.29296875],<br>  [ 1.234375    0.4873047  -2.3203125  ...  0.58447266 -0.67529297 -1.0273438 ],<br>  ```[ 0.25585938  0.14770508 -0.23474121 ... -0.9873047  -0.7993164  -1.3955078 ]```,<br>  ...<br>  ```[ 1.3798828   0.7055664  -1.1689453  ... -0.06005859 -0.79052734 -0.65722656]```,<br>  ```[ 1.1875      0.51660156 -1.1699219  ... -0.0579834  -0.74560547 -0.7236328 ]```,<br>  ```[ 1.1738281   0.54248047 -1.6289062  ... -0.23840332 -0.82128906 -0.9790039 ]]]``` | [[[ 1.9384766   0.09979248 -2.140625   ...  0.14038086 -1.3173828 -0.2939453 ],<br>  [ 1.2363281   0.48657227 -2.3203125  ...  0.5839844  -0.6767578 -1.0263672 ],<br>  ```[ 1.3183594   0.22973633 -2.1152344  ...  0.32666016 -0.8300781 -1.2978516 ]```,<br>  ...<br>  ```[ 1.5742188   2.6425781  -3.5605469  ...  1.3398438   0.6098633   -0.45166016]```,<br>  ```[ 1.6054688   2.7617188  -3.4492188  ...  1.4746094   0.49902344  -0.34106445]```,<br>  ```[ 1.8447266   2.71875    -4.015625   ...  1.5742188   0.71728516  0.30981445]]]``` |

Here, ```pad_token_id = 242834```. 

Could you help us understand how to ensure output parity between A100 and H100? Specifically, we aim to match the A100 output on H100. Any insights into the cause of this discrepancy and possible solutions would be greatly appreciated.



### To reproduce

Test code:

```python 
import onnxruntime as ort
import numpy as np

def main():
    # Path to the ONNX model
    model_path = "model.onnx"

    # Create an ONNX Runtime session with CUDA execution, explicitly setting inter_op_num_threads
    session_options = ort.SessionOptions()
    session_options.inter_op_num_threads = 1
    session_options.intra_op_num_threads = 1
    session = ort.InferenceSession(model_path, sess_options=session_options, providers=["CUDAExecutionProvider"])

    # Sample input batch
    input_batch = {
        'src_tokens': np.array([[31, 218232, 242834, 242834, 242834, 242834, 242834, 242834,
                                 242834, 242834, 242834, 242834, 242834, 242834, 242834, 242834]], dtype=np.int64),
        'attention_mask': np.array([[1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=np.int32)    }

    # Run inference
    print ("Input: ", input_batch)
    outputs = session.run(None, input_batch)
    print ("Output: ", outputs[0])

if __name__ == "__main__":
    main()
```

Test model: [Link](https://drive.google.com/drive/folders/1VMNOnAJi-V-9EQCi_mhcVqGBXjT3eC8U?usp=sharing)


### Urgency

We are using ONNX Runtime for model inference in Microsoft Ads. This issue is blocking our deployments as we are observing inconsistent results between two GPUs (A100 vs. H100)

### Platform

Linux

### OS Version

Ubuntu 22.04.3 LTS

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.19.0

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 12.2, CuDNN 9.7.1

**Comment by xadupre** [comment]
In the todo list, I would check first if this happens without onnxruntime optimization first. Then I would try to find the kernel responsible for the discrepancies. I can be done by changing the output of the model to match the ones of a node (see https://onnx.ai/onnx/api/utils.html#onnx.utils.extract_model). It must be done for every node until this introducting the error is found.

**Comment by tianleiwu** [comment]
The attention mask `[[1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]` means only first two tokens are valid (remaining tokens are padding). The output for padding tokens do not impact end to end result (you may drop those padding tokens in the end). You can see that hidden state for first two tokens have parity (slight difference is expected in different GPUs):

![Image](https://github.com/user-attachments/assets/04cf3346-5a9c-4ba9-b2da-58bb91ba848f)

**Comment by akashkm99** [comment]
Hi @tianleiwu, thanks for the detailed response!

In our scenario, the model is explicitly trained to utilize hidden representations from padding tokens. Since it's a non-autoregressive generative model that produces predictions at padding positions as well, differences in padding token representations impacts the final output.

Given this, is there currently a workaround (e.g., specific environment variables or configuration flags) to ensure the H100 padding token representations closely match the current A100 outputs before the next release? Additionally, could you please share the tentative timeline for this upcoming release?

Thanks again!

**Comment by tianleiwu** [comment]
@akashkm99, please try 
```
providers = [("CUDAExecutionProvider", {"sdpa_kernel": 1})]
sess_options = ort.SessionOptions()
sess = ort.InferenceSession("bert.onnx", sess_options=sess_options, providers=providers)
```

There are different options for sdpa_kernel, see https://github.com/microsoft/onnxruntime/blob/154e3b7d3da84d38c2cf7e28115e6dfb2e563a5f/onnxruntime/contrib_ops/cpu/bert/attention_common.h#L66-L81
Note that if the kernel is not available, it will fall back to unfused kernel. 

Next release is about 3 months.

**Comment by snnn** [comment]
This issue has been automatically closed because it has been marked as 'stale' for more than 30 days without activity. If you believe this is still an issue, please feel free to reopen it.

**Comment by jianlany** [comment]
Hello, I would like to follow up on this issue.

It has been 5 months since the issue opened. I checked recent release notes and did not find anything related to the TRT_FUSED_ATTENTION being supported by H100. Can you clarify if it has been added?

Thank you!
Jianlan
